### PR TITLE
Native notifications

### PR DESCRIPTION
**Pull request by vivangkumar**
This is just a v1. Introduces support for sending native notifications via Pusher.

**DO NOT MERGE YET**

TODO: ~~Change the endpoint once finalised~~

### CODE REVIEW

**Comment by jpatel531** [comment]
@vivangkumar - can you validate that only one interest is sent?

**Comment by jpatel531** [comment]
also for dev experience, you can duplicate what's in platforms/:platform/client.go and validate it in the library

**Comment by vivangkumar** [comment]
I thought we settled on sending an array of interests?

**Comment by jpatel531** [comment]
yes but atm the server only supports sending to one interest

File: lib/pusher/native_notification/client.rb [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      end
+
+      # Validate payload
+      # `time_to_live` -> value b/w 0 and 241920
+      # If the `notification` key is provided, ensure
+      # that there is an accompanying `title` and `icon`
+      # field
+      def validate_payload(payload)
+        # Restricted keys
+        RESTRICTED_GCM_PAYLOAD_KEYS.each { |k| payload.delete(k) }
```
Comment (jpatel531) [inline lib/pusher/native_notification/client.rb]: @vivangkumar you need do to it on `payload[:gcm]` and if it exists

Comment (vivangkumar) [inline lib/pusher/native_notification/client.rb]: Yikes. Thanks

File: lib/pusher/native_notification/client.rb [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      # that there is an accompanying `title` and `icon`
+      # field
+      def validate_payload(payload)
+        unless (payload.has_key?(:apns) || payload.has_key?(:gcm))
+          raise Pusher::Error, "GCM or APNS data must be provided"
+        end
+
+        if (gcm_payload = payload[:gcm])
+          # Restricted keys
+          RESTRICTED_GCM_PAYLOAD_KEYS.each { |k| gcm_payload.delete(k) }
```
Comment (zimbatm) [inline lib/pusher/native_notification/client.rb]: Shouldn't that be filtered on the server side anyways?

File: lib/pusher/native_notification/client.rb [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+          if (ttl = gcm_payload[:time_to_live])
+
+            if ttl.to_i < 0 || ttl.to_i > GCM_TTL
+              raise Pusher::Error, "Time to live must be between 0 and 241920 (4 weeks)"
```
Comment (zimbatm) [inline lib/pusher/native_notification/client.rb]: This also should be an error returned by the server

Comment (vivangkumar) [inline lib/pusher/native_notification/client.rb]: There is validation on the server for all cases. Probably better to validate both places?

**Comment by vivangkumar** [comment]
@zimbatm im happy with this as a v1, however not sure why the builds break. Webmock seems to complain as it doesn't seem to recognise the stubs :/

**Pull request merged**